js/modules/Monaco/MonacoEditor.js

```diff
@@ -0,0 +1,125 @@
+/**
+ * ---------------------------------------------------------------------
+ *
+ * GLPI - Gestionnaire Libre de Parc Informatique
+ *
+ * http://glpi-project.org
+ *
+ * @copyright 2015-2023 Teclib' and contributors.
+ * @copyright 2003-2014 by the INDEPNET Development Team.
+ * @licence   https://www.gnu.org/licenses/gpl-3.0.html
+ *
+ * ---------------------------------------------------------------------
+ *
+ * LICENSE
+ *
+ * This file is part of GLPI.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+window.GLPI = window.GLPI || {};
+
+/**
+ * @typedef CompletionItemDefinition
+ * @property {string} name The name of the completion item.
+ * @property {string} type The type of the completion item. This corresponds to a type in the {@link CompletionItemKind} enum.
+ */
+export default class MonacoEditor {
+    /**
+     *
+     * @param {string} element_id The ID of the DIV to create the editor in
+     * @param {string} language The code language
+     * @param {string} value The default value for the editor
+     * @param {CompletionItemDefinition[]} completions List of completion items
+     */
+    constructor(element_id, language, value = '', completions = []) {
+        const el = document.getElementById(element_id);
+        const trigger_characters = {
+            twig: ['{', ' '],
+        };
+        window.monaco.languages.registerCompletionItemProvider(language, {
```

Comment:
I'm not sure it is worth it at this time.
Completion data is globally managed in the `window.monaco` object which affects any editor created from it.
If plugins really want to modify this, they have access to the window global. Having it done as methods on the MonacoEditor class could confuse them into thinking they are only changing the behavior of that one editor.
Also, you can have multiple providers at a time. So, if you really want to replace the provider, you need to track all the Disposable items returned by registerCompletionItemProvider.

I don't see a need for a plugin to control the behavior of the trigger characters or provider. They do have the option of passing custom completion data (variables, functions, etc) per-editor though.

Reply:
> I don't see a need for a plugin to control the behavior of the trigger characters or provider.

To clarify, we control the available languages on the GLPI side because it is in the webpack config. So, if a language is allowed from the core, we can provide and maintain a set of good providers for them too (if the default isn't sufficient).